doc/developer/design/20230225_compute_introspection_schema.md

```diff
@@ -0,0 +1,217 @@
+GROUP BY operator_id, worker_id"
+```
+
+This dimension is applied inconsistently: not all relations that
+report a count or sum in the data are derived from a corresponding
+`_internal` source. 
+
+`raw` variants, on the other hand, measure some time value in
+nanoseconds, whereas their corresponding non-`raw` views convert the
+times to the SQL interval type.
+
+Confusingly, not all times are reported as intervals; for example,
+`mz_scheduling_parks` has `slept_for` and `requested` columns, whose
+type is `bigint` and whose unit is undocumented (in fact, it is nanoseconds).
+
+#### Inconsistent presence of `sum` column in histograms
+
+Some histograms include a per-bucket sum, which allows consumers to
+compute average or total values precisely. Confusingly, not all
+histograms report such a column (e.g. `mz_worker_compute_delays` does
+not, whereas `mz_worker_raw_compute_delays` does).
+
+#### Lack of clarity around per-replica vs. global relations
+
+Many of the relations are logging sources that run locally in
+clusters, are not persisted, and report different data depending on
+the currently selected replica. Views based on these sources are also
+per-replica. It is not straightforward to tell from our docs which is
+the case for each relation.
+
+Furthermore, sometimes the same logical value is reported by two
+relations, one reporting `clusterd`'s view of the world, and the other
+reporting `environmentd`'s. For example, `mz_compute_frontiers`
+vs. `mz_cluster_replica_frontiers`. However, the names do not make it
+clear what the distinction is. 
+
+### Suggested fixes for schema inconsistencies
+
+* All per-worker relations should have a corresponding "global" view
+  where the results are aggregated across workers in the way most
+  likely to be useful (for example, summing, taking the minimum, or
+  restricting to worker 0). Because these global views are more likely
+  to be useful than the underlying sources, they should have the basic
+  name (e.g., `mz_arrangement_sharing`). The underlying source should
+  have a name with `per_worker` suffixed (e.g.,
+  `mz_arrangement_sharing_per_worker`). Furthermore, the per-worker
+  variant should be clearly less prominent in documentation (either at
+  the bottom of the page, or in a "see more" section that has to be
+  clicked to expand).
+* The expectations about per-worker relations should be clearly
+  documented (e.g., we should clearly say whether the values for each
+  worker are expected to be the same during normal operation).
+* All relations that report a simple count or sum should be views based on
+  an `_internal` variant that maintains the count or sum in the
+  Differential diff field. We may consider renaming this, since it
+  conflicts confusingly with the overall name `mz_internal`.
+* `raw` variants should be removed, and the underlying dataflows
+  should be in terms of intervals.
+* All histogram views or sources should have a per-bucket `sum`
+  column.
```

Comment:
We discussed offline the option to instead drop the `sum` field from all histograms. This is a summary of that discussion and a bit of research on my side.

* We introduced the `sum` field [because Prometheus expects it](https://prometheus.io/docs/practices/histograms/#count-and-sum-of-observations).
* The use of this field is that you can compute averages (both per bucket and across buckets).
* We can compute approximate the sum using the bucket upper-bound `b`:
    * `b * count` is an over-approximation. For our power-of-2 based histograms it is at worst off by a factor of 2.
    *  If we now the upper bound of the previous bucket `b'`, we could get a better approximation with `((b + b') / 2) * count`
* Setting the sum to 0 or leaving it out when reporting a histogram to Prometheus will make the resulting metric show 0. While it doesn't break anything, it will produce confusing results if one tries to compute an average.

So there are two viable options we have to decide between: Adding a sum field (or multiple) to each histogram source, or not doing that and approximating the sum when we need it.

I think I prefer the second option (approximating the sum) because:
* We don't know if we (or our users) need accurate averages.
* Not having the sum field means having less code, and possibly less resource usage for maintaining it.
* We can always add the field later if we find a strong use-case. Removing it later would be more disruptive (even for our unstable introspection sources).

Reply:
I agree with @teskje here. The `sum` field is not so important that we need to keep it if it is introducing difficulties.

Note that we will have to modify the prometheus-sql-exporter in the cloud repo to continue understanding the histograms if we remove the sum field